File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/telemetry/TelemetrySink.java
package org.hyperledger.iroha.android.telemetry;

import java.util.Map;
import org.hyperledger.iroha.android.client.ClientResponse;

/**
 * Sink that consumes telemetry events emitted by {@link org.hyperledger.iroha.android.client.ClientObserver}
 * implementations.
 *
 * <p>Mobile apps can supply an implementation backed by their preferred observability stack (e.g.,
 * OpenTelemetry, structured logs, or Norito exporters).
 */
public interface TelemetrySink {

  void onRequest(TelemetryRecord record);

  void onResponse(TelemetryRecord record, ClientResponse response);

  void onFailure(TelemetryRecord record, Throwable error);

  /**
   * Emits structured telemetry signals that are not tied to HTTP lifecycle callbacks (e.g., queue
   * depth gauges). Default implementation is a no-op so existing sinks remain compatible.
   */
  default void emitSignal(final String signalId, final Map<String, Object> fields) {
    // Intentionally empty.
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/telemetry/TelemetrySink.kt
package org.hyperledger.iroha.sdk.telemetry

import org.hyperledger.iroha.sdk.client.ClientResponse

/**
 * Sink that consumes telemetry events emitted by `ClientObserver`
 * implementations.
 *
 * Mobile apps can supply an implementation backed by their preferred observability stack (e.g.,
 * OpenTelemetry, structured logs, or Norito exporters).
 */
interface TelemetrySink {

    fun onRequest(record: TelemetryRecord)

    fun onResponse(record: TelemetryRecord, response: ClientResponse)

    fun onFailure(record: TelemetryRecord, error: Throwable)

    /**
     * Emits structured telemetry signals that are not tied to HTTP lifecycle callbacks (e.g., queue
     * depth gauges). Default implementation is a no-op so existing sinks remain compatible.
     */
    fun emitSignal(signalId: String, fields: Map<String, Any>) {
        // Intentionally empty.
    }
}
